### ISSUE TITLE
swin vit transfigure

### ISSUE THREAD

**Issue by DevranOz** [DevranOz/devran#1]
Alz VİT RANSFİGURE SWIN

from google.colab import drive
drive.mount('/content/drive')

!pip install timm grad-cam

import os
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset
from torchvision import transforms

import timm
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image

class AlzheimerDataset(Dataset):
    def __init__(self, root_dir, transform=None):
        self.image_paths = []
        self.labels = []
        self.transform = transform
        classes = sorted(os.listdir(root_dir))
        self.class_to_idx = {cls_name: idx for idx, cls_name in enumerate(classes)}

        for cls in classes:
            cls_dir = os.path.join(root_dir, cls)
            for img_name in os.listdir(cls_dir):
                img_path = os.path.join(cls_dir, img_name)
                self.image_paths.append(img_path)
                self.labels.append(self.class_to_idx[cls])

    def __len__(self):
        return len(self.image_paths)

    def __getitem__(self, idx):
        image = Image.open(self.image_paths[idx]).convert('RGB')
        label = self.labels[idx]
        if self.transform:
            image = self.transform(image)
        return image, label

train_dir = "/content/drive/MyDrive/Alz/Train"
test_dir = "/content/drive/MyDrive/Alz/Test"

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(),
    transforms.RandomRotation(15),
    transforms.ColorJitter(brightness=0.2, contrast=0.2),
    transforms.ToTensor(),
    transforms.Normalize([0.5]*3, [0.5]*3)
])

train_dataset = AlzheimerDataset(train_dir, transform)
test_dataset = AlzheimerDataset(test_dir, transform)

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=16, shuffle=False)

classes = sorted(os.listdir(train_dir))
num_classes = len(classes)

models_dict = {
    "ViT": "vit_base_patch16_224",
    "Swin": "swin_base_patch4_window7_224",
    "TransFigure": "convnext_base"
}

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
results = {}

from google.colab import drive
drive.mount('/content/drive')
!pip install timm grad-cam

import os
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, Dataset
from torchvision import transforms

import timm
from pytorch_grad_cam import GradCAM
from pytorch_grad_cam.utils.image import show_cam_on_image

train_dir = "/content/drive/MyDrive/Alz/Train"
test_dir = "/content/drive/MyDrive/Alz/Test"

transform = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(),
    transforms.RandomRotation(15),
    transforms.ColorJitter(brightness=0.2, contrast=0.2),
    transforms.ToTensor(),
    transforms.Normalize([0.5]*3, [0.5]*3)
])

train_dataset = AlzheimerDataset(train_dir, transform)
test_dataset = AlzheimerDataset(test_dir, transform)

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=16, shuffle=False)

classes = sorted(os.listdir(train_dir))
num_classes = len(classes)

for model_name, model_arch in models_dict.items():
    print(f"\nTraining: {model_name}")

    # Modeli yükle
    model = timm.create_model(model_arch, pretrained=True, num_classes=num_classes)
    model.to(device)

    # Kayıp fonksiyonu ve optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)

    train_acc_history = []
    test_acc_history = []

    # EPOCH DÖNGÜSÜ
    for epoch in range(40):
        model.train()
        correct = 0
        total = 0

        for images, labels in tqdm(train_loader, desc=f"{model_name} - Epoch {epoch+1}"):
            images, labels = images.to(device), labels.to(device)
            outputs = model(images)
            loss = criterion(outputs, labels)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            _, predicted = outputs.max(1)
            total += labels.size(0)
            correct += predicted.eq(labels).sum().item()

        train_acc = 100 * correct / total
        train_acc_history.append(train_acc)

        # TEST
        model.eval()
        correct = 0
        total = 0
        with torch.no_grad():
            for images, labels in test_loader:
                images, labels = images.to(device), labels.to(device)
                outputs = model(images)
                _, predicted = outputs.max(1)
                total += labels.size(0)
                correct += predicted.eq(labels).sum().item()
        test_acc = 100 * correct / total
        test_acc_history.append(test_acc)

        print(f"Epoch {epoch+1}: Train Acc = {train_acc:.2f}%, Test Acc = {test_acc:.2f}%")

    results[model_name] = (train_acc_history, test_acc_history)

    # Grad-CAM AÇIKLAMA
    image, label = test_dataset[0]
    input_tensor = image.unsqueeze(0).to(device)

    # Grad-CAM için geçerli katmanlar
    if "vit" in model_arch:
        target_layers = [model.patch_embed.proj]  # Conv2d
    elif "swin" in model_arch:
        try:
            target_layers = [model.features[-1].blocks[-1].norm1]
        except:
            print("[Uyarı] Swin modelde özellik yapısı tanınamadı, varsayılan katman kullanılacak.")
            target_layers = [model.norm]
    else:  # ConvNeXt (TransFigure)
        try:
            last_stage = model.stages[-1]
            last_block = list(last_stage.children())[-1]  # son blok (örneğin ConvNeXtBlock)
            target_layers = [last_block.dwconv]  # son depthwise conv
        except Exception as e:
            print(f"[HATA] ConvNeXt katmanı seçilirken sorun oluştu: {e}")
            continue

    # GradCAM çalıştır
    model.eval()
    cam = GradCAM(model=model, target_layers=target_layers)

    try:
        grayscale_cam = cam(input_tensor=input_tensor, targets=None)[0, :]

        # Görüntüyü normalize et ve CAM görseli oluştur
        img = image.permute(1, 2, 0).cpu().numpy()
        img = (img - img.min()) / (img.max() - img.min())
        vis = show_cam_on_image(img, grayscale_cam, use_rgb=True)

        # Göster
        plt.imshow(vis)
        plt.title(f"{model_name} Grad-CAM")
        plt.axis("off")
        plt.show()

    except Exception as e:
        print(f"[HATA - Grad-CAM] {model_name} için CAM üretilemedi: {e}")

# ---  --- Doğruluk Grafiği 
plt.figure(figsize=(10, 5))
for model_name, (train_acc, test_acc) in results.items():
    if len(train_acc) > 0 and len(test_acc) > 0:
        plt.plot(train_acc, label=f"{model_name} Train")
        plt.plot(test_acc, label=f"{model_name} Test", linestyle="--")
plt.title("Model Training and Testing Accuracy")
plt.xlabel("Epoch")
plt.ylabel("Accuracy (%)")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()
